src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -716,25 +729,35 @@ protected virtual void Dispose(bool finalize)
         _workQueueTokenSource.Cancel();
     }
 
-    private async ValueTask ProcessWorkQueueAsync(ImmutableSegmentedList<Action> list, CancellationToken cancellationToken)
+    private async ValueTask ProcessEventHandlerWorkQueueAsync(ImmutableSegmentedList<(EventArgs Args, EventHandlerSet handlerSet)> list, CancellationToken cancellationToken)
     {
-        // Hop over to the right scheduler to execute all this work.
+        ProcessWorkQueueHelper(list, shouldRaise: static options => !options.RequiresMainThread, cancellationToken);
+
         await Task.Factory.StartNew(() =>
         {
-            foreach (var item in list)
+            ProcessWorkQueueHelper(list, shouldRaise: static options => options.RequiresMainThread, cancellationToken);
+        }, cancellationToken, TaskCreationOptions.None, _taskSchedulerProvider.CurrentContextScheduler).ConfigureAwait(false);
```

Comment:
this hop feels undesirable if there are no items in the list that require the main thread.